### PR TITLE
Async usage analyzers

### PR DESCRIPTION
**Pull request by leonskim**
@kevinvangelder Please review this, sir 👍 

- Added `AsyncUsageAnalyzers` and fixed warnings
- Otherwise just suppressed the false positives

### CODE REVIEW

**Comment by rozele** [comment]
Might make sense to do this for ReactNative.csproj (for UWP) as well

File: ReactWindows/ReactNative.Shared/Bridge/Queue/MessageQueueThread.cs [hunk 1]
```diff
@@ -281,7 +281,10 @@ public AnyBackgroundMessageQueueThread(string name, Action<Exception> handler)
                 _taskFactory = new TaskFactory(_taskScheduler);
             }
 
+            [System.Diagnostics.CodeAnalysis.SuppressMessage("AsyncUsage.CSharp.Reliability", "AvoidAsyncVoid", Justification = "Reviewed.")]
```
Comment (leonskim) [inline ReactWindows/ReactNative.Shared/Bridge/Queue/MessageQueueThread.cs]: `#pragma warning disable/enable` is to suppress violations for false positives, and `SuppressMessage` is an optional code that specifies the analysis tool, identifier and so on.

**Comment by leonskim** [comment]
Will do add it for UWP :+1:

File: ReactWindows/ReactNative.Net46/Bridge/DispatcherHelpers.cs [hunk 2]
```diff
@@ -54,14 +54,17 @@ public static bool IsOnDispatcher()
...
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("AsyncUsage.CSharp.Reliability", "AvoidAsyncVoid", Justification = "Reviewed.")]
+#pragma warning disable AvoidAsyncVoid
         public static async void RunOnDispatcher(Action action)
+#pragma warning restore AvoidAsyncVoid
         {
             AssertDispatcherSet();
 
             await CurrentDispatcher.InvokeAsync(action).Task.ConfigureAwait(false);
```
Comment (kevinvangelder) [inline ReactWindows/ReactNative.Net46/Bridge/DispatcherHelpers.cs]: Doesn't this return a Task?

Comment (leonskim) [inline ReactWindows/ReactNative.Net46/Bridge/DispatcherHelpers.cs]: @kevinvangelder It turns unaccessible from other classes when I change the return type and  add suffix. I'll leave it for now and check if there are same issues on UWP.

Comment (kevinvangelder) [inline ReactWindows/ReactNative.Net46/Bridge/DispatcherHelpers.cs]: Okay.

File: ReactWindows/ReactNative.Net46/Views/Image/ReactImageManager.cs [hunk 3]
```diff
@@ -274,7 +274,10 @@ private void OnImageStatusUpdate(Border view, ImageStatusEventData status)
         /// </summary>
         /// <param name="view">The image view instance.</param>
         /// <param name="source">The source URI.</param>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("AsyncUsage.CSharp.Reliability", "AvoidAsyncVoid", Justification = "Reviewed.")]
+#pragma warning disable AvoidAsyncVoid
         private async void SetUriFromSingleSource(Border view, string source)
```
Comment (leonskim) [inline ReactWindows/ReactNative.Net46/Views/Image/ReactImageManager.cs]: Hey @rozele, I suppressed `AvoidAsyncVoid` on `SetUriFromSingleSource`, but would we want this method to be `async` without `await` operator?

Comment (kevinvangelder) [inline ReactWindows/ReactNative.Net46/Views/Image/ReactImageManager.cs]: @leonskim Yes, if there is no `await` inside then the `async` is unnecessary, I probably just forgot to remove it when I translated the file.

File: ReactWindows/ReactNative/Bridge/DispatcherHelpers.cs [hunk 4]
```diff
@@ -20,16 +20,16 @@ public static bool IsOnDispatcher()
             return CoreWindow.GetForCurrentThread()?.Dispatcher != null;
         }
 
-        public static async void RunOnDispatcher(DispatchedHandler action)
+        public static async Task RunOnDispatcherAsync(DispatchedHandler action)
```
Comment (rozele) [inline ReactWindows/ReactNative/Bridge/DispatcherHelpers.cs]: >Task [](start = 28, length = 4)

Why does this need to return Task?  This is intended to be fire and forget to queue up a task on the dispatcher.  If they want to await the response, that's what CallOnDispatcher is for (now CallOnDispatcherAsync, which is fine)

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs [hunk 5]
```diff
@@ -217,11 +217,14 @@ private string BuildModulesConfig()
             }
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("AsyncUsage.CSharp.Reliability", "AvoidAsyncVoid", Justification = "Reviewed.")]
         private void HandleException(Exception ex)
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: Do you need this SuppressMessage here?  It seems that the warning is for the delegate below.

**Comment by rozele** [comment]
:clock1:

**Comment by leonskim** [comment]
Thanks for the review @rozele, I'll fix those later tonight so you can see it tomorrow :+1:

File: ReactWindows/ReactNative/Modules/StatusBar/IStatusBar.cs [hunk 6]
```diff
@@ -24,7 +24,7 @@ public interface IStatusBar
         /// Hides the status bar.
         /// </summary>
         /// <returns>The asynchronous results of the operation. Use this to determine when the async call is complete</returns>
-        [System.Diagnostics.CodeAnalysis.SuppressMessage("AsyncUsage.CSharp.Reliability", "AvoidAsyncSuffix", Justification = "Reviewed.")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("AsyncUsage.CSharp.Reliability", "AvoidAsyncSuffix", Justification = "Intended name.")]
 #pragma warning disable AvoidAsyncSuffix 
         IAsyncAction HideAsync();
```
Comment (matthargett) [inline ReactWindows/ReactNative/Modules/StatusBar/IStatusBar.cs]: Really irritating that this attribute is marked as Inherited=false :( Is there some other attribute that inherits from this one and is inheritable, @rozele ?

File: ReactWindows/ReactNative/Modules/Dialog/DialogModule.cs [hunk 7]
```diff
@@ -67,8 +71,11 @@ public void OnDestroy()
         {
         }
 
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("AsyncUsage.CSharp.Reliability", "AvoidAsyncVoid", Justification = "React method must return void.")]
         [ReactMethod]
-        public void showAlert(
+#pragma warning disable AvoidAsyncVoid
+        public async void showAlert(
```
Comment (rozele) [inline ReactWindows/ReactNative/Modules/Dialog/DialogModule.cs]: >async [](start = 15, length = 5)

No await here.

**Comment by rozele** [comment]
Please review for unnecessary async usages where there are no awaits.  There seems to be a few after reverting RunOnDispatcher from async.

**Comment by rozele** [comment]
Closing and reopening to trigger CLA bot

**Comment by msftgits** [comment]
Hi, I am closing and re-opening this PR to bump the CLA bot. Sorry for the inconvenience!

**Comment by rozele** [comment]
This is pretty far out of date. We'll need another pass on CodeAnalysis / Roslyn Analyzers at some point.

**Pull request closed without merging**